### PR TITLE
A few more regex test

### PR DESCRIPTION
**Pull request by danmoseley**
Borrow a few parser tests from Rust (https://github.com/rust-lang/regex). I'll add some execution tests later, although unfortunately many we already have as they derive from the AT&T tests.

Work around #56945 which only seems to affect one of my machines.

### CODE REVIEW

**Review by stephentoub — APPROVED** [review]

File: src/libraries/System.IO.FileSystem/tests/File/EncryptDecrypt.cs [hunk 1]
```diff
@@ -43,9 +43,11 @@ public static void EncryptDecrypt_Read()
                 {
                     File.Encrypt(tmpFileName);
                 }
-                catch (IOException e) when (e.HResult == unchecked((int)0x80070490))
+                catch (IOException e) when (e.HResult == unchecked((int)0x80070490) ||
+                                           (e.HResult == unchecked((int)0x80071776)))
```
Comment (pgovind) [inline src/libraries/System.IO.FileSystem/tests/File/EncryptDecrypt.cs:47]: I'm guessing this is the workaround you called out in the PR description? Do we want to merge this in? Or did you have it here just to make CI pass?

**Review by pgovind — APPROVED** [review]
Nothing more to add after Stephen's comments. LGTM!

File: src/libraries/System.Text.RegularExpressions/tests/RegexParserTests.netfx.cs [hunk 2]
```diff
@@ -28,11 +28,32 @@ public partial class RegexParserTests
...
+                action();
+            }
+            catch (ArgumentException)
+            {
+                return;
+            }
+            catch (Exception e)
+            {
+                throw new XunitException($"Expected RegexParseException or no exception -> Actual: ({e})");
+            }
```
[the remark this answers is not in the document]
Comment (danmoseley) [reply src/libraries/System.Text.RegularExpressions/tests/RegexParserTests.netfx.cs:56]: ooh I didn't know about Record.Exception.

File: src/libraries/System.IO.FileSystem/tests/File/EncryptDecrypt.cs (hunk 1 above)
Comment (danmoseley) [reply src/libraries/System.IO.FileSystem/tests/File/EncryptDecrypt.cs:47]: Yes, I would like to merge it in. It didn't seem worth its own PR.

File: src/libraries/System.Text.RegularExpressions/tests/RegexParserTests.cs [hunk 3]
```diff
@@ -716,67 +701,66 @@ public void Parse(string pattern, RegexOptions options, object errorObj, int off
...
-                ParseSubTree(pattern.Substring(0, i) + pattern.Substring(i + 1), options);
-            }
-        }
+            RegexParseError? error = (RegexParseError?)errorObj;
 
-        static void ParseTree(string pattern, RegexOptions options, RegexParseError? error, int offset)
-        {
             if (error != null)
             {
+                Assert.True(error.Value >= 0);
```
[the remark this answers is not in the document]
Comment (danmoseley) [reply src/libraries/System.Text.RegularExpressions/tests/RegexParserTests.cs]: (It's InRange)

**Comment by danmoseley** [comment]
@steveisok I have seen this MacCatalyst was canceled on other PR's.

I'm merging because the possibility of these test changes having an issue specific to MacCatalyst is ~zero.

**Pull request merged**